src/OrchardCore/OrchardCore.Modules/ModularTenantContainerMiddleware.cs

```diff
@@ -76,11 +78,27 @@ public class ModularTenantContainerMiddleware
                         finally
                         {                            
                             semaphore.Release();
-                            _semaphores.TryRemove(shellSetting.Name, out semaphore);
+                            _semaphores.TryRemove(shellSettings.Name, out semaphore);
                         }
                     }
                     
                     await _next.Invoke(httpContext);
+                    deferredTaskEngine = scope.ServiceProvider.GetService<IDeferredTaskEngine>();
+                }
+
+                // Create a new scope only if there are pending tasks
+                if (deferredTaskEngine != null && deferredTaskEngine.HasPendingTasks)
+                {
+                    shellContext = _orchardHost.GetOrCreateShellContext(shellSettings);
+
+                    if (!shellContext.Released)
```

Comment:
Here you test for `Released` which can make sense. But we don't do it in the previous block, so it's inconsistent. I agree there might be an issue here, between the two calls. So we need a method that will ensure that it's create a shell and returning a tracked scope atomically instead.